activerecord/lib/active_record/relation.rb

```diff
@@ -315,9 +315,15 @@ def scoping
       klass.current_scope = previous
     end
 
-    def _exec_scope(*args, &block) # :nodoc:
+    def _exec_scope(name, *args, &block) # :nodoc:
       @delegate_to_klass = true
-      instance_exec(*args, &block) || self
+      instance_exec(*args, &block).tap do |result|
+        valid_scope_result = result.is_a?(Relation) || !result
+        unless valid_scope_result
+          logger&.warn "#{klass}.#{name} returned a value of type #{result.class}. " \
+            "Scopes should return ActiveRecord::Relation, nil or false!"
```

Comment:
We should probably mention that a regular class method is preferable in this case:

> Scopes are intended to return ActiveRecord::Relation objects; consider defining a class method instead.

Reply:
I like this.👌